src/macros.rs

```diff
@@ -1,3 +1,39 @@
+/// Create an [`IndexMap`][crate::IndexMap] from a list of key-value pairs and a custom hasher
+///
+/// ## Example
+///
+/// ```
+/// use indexmap::indexmap_with_hasher;
+/// use std::hash::DefaultHasher;
+///
+/// let map = indexmap_with_hasher!{
+///     DefaultHasher;
+///     "a" => 1,
+///     "b" => 2,
+/// };
+/// assert_eq!(map["a"], 1);
+/// assert_eq!(map["b"], 2);
+/// assert_eq!(map.get("c"), None);
+///
+/// // "a" is the first key
+/// assert_eq!(map.keys().next(), Some(&"a"));
+/// ```
+#[macro_export]
+macro_rules! indexmap_with_hasher {
+    ($S:ty; $($key:expr => $value:expr,)+) => { $crate::indexmap_with_hasher!($S; $($key => $value),+) };
+    ($S:ty; $($key:expr => $value:expr),*) => {{
+        type BuildS = core::hash::BuildHasherDefault<$S>;
+        const CAP: usize = <[()]>::len(&[$({ stringify!($key); }),*]);
+        #[allow(unused_mut)]
+        // Specify your custom S (must implement Default + Hasher) as the hasher:
+        let mut map = $crate::IndexMap::<_, _, BuildS>::with_capacity_and_hasher(CAP, <BuildS>::default());
```

Comment:
I don't think we need the map turbofish or type alias at all. This `<BuildS>::default()` is just giving you a `BuildS` (which is really `S`), the same as its `new()` method. So I think this can be more like:
```suggestion
        let builder = ::core::hash::BuildHasherDefault::<$H>::new();
        let mut map = $crate::IndexMap::with_capacity_and_hasher(CAP, builder);
```